crates/turbo-tasks-memory/src/memory_backend.rs

```diff
@@ -261,6 +275,97 @@ impl MemoryBackend {
             }
         }
     }
+
+    pub(crate) fn get_or_create_read_collectibles_task(
+        &self,
+        scope_id: TaskScopeId,
+        trait_type: TraitTypeId,
+        parent_task: TaskId,
+        root_scoped: bool,
+        turbo_tasks: &dyn TurboTasksBackendApi,
+    ) -> TaskId {
+        if let Some(task) = self.lookup_and_connect_task(
+            parent_task,
+            &self.read_collectibles_task_cache,
+            &(scope_id, trait_type),
+            turbo_tasks,
+        ) {
+            // fast pass without creating a new task
+            task
+        } else {
+            // slow pass with key lock
+            let id = turbo_tasks.get_fresh_task_id();
+            let task =
+                Task::new_read_collectibles(id, scope_id, trait_type, turbo_tasks.stats_type());
+            // Safety: We have a fresh task id that nobody knows about yet
```

Comment:
This could perhaps be ensured via a newtype `FreshTaskId`/`FreshTask`. Only once you `insert` the task do you get a `Task` with a `TaskId`. `remove`ing the task gives you back a `FreshTaskId`.

Reply:
That's a cool idea...